### PR TITLE
Adds logging PennyBOT logs to Discord

### PR DESCRIPTION
**Pull request by MahdiBM**

### CODE REVIEW

File: CODE/task-def.json [hunk 1]
```diff
@@ -34,6 +34,10 @@
                 {
                     "valueFrom": "arn:aws:secretsmanager:eu-west-1:177420307256:secret:prod/penny/penny-bot/discord-bot-token-VmhBXH",
                     "name": "BOT_TOKEN"
+                },
+                {
+                    "valueFrom": "arn:aws:secretsmanager:eu-west-1:177420307256:secret:prod/penny/penny-bot/discord-bot-token-VmhBXH",
```
Comment (MahdiBM) [inline CODE/task-def.json:39]: I'm also not sure if we don't need to change this `valueFrom` key or not.

File: CODE/Sources/PennyBOT/Constants.swift [hunk 2]
```diff
@@ -3,8 +3,10 @@ import DiscordBM
 
 enum Constants {
     static let internalChannelId = "441327731486097429"
+    static let botDevUserId = "290483761559240704"
     static let botToken: String! = ProcessInfo.processInfo.environment["BOT_TOKEN"]
     static var botId: String! = ProcessInfo.processInfo.environment["BOT_APP_ID"]
+    static var loggingWebhookUrl: String! = ProcessInfo.processInfo.environment["LOGGING_WEBHOOK_URL"]
```
Comment (MahdiBM) [inline CODE/Sources/PennyBOT/Constants.swift:9]: Need to add `LOGGING_WEBHOOK_URL` to aws.
First need to create a webhook in Discord, of course.

File: CODE/Sources/PennyBOT/Constants.swift [hunk 3]
```diff
@@ -3,8 +3,10 @@ import DiscordBM
 
 enum Constants {
     static let internalChannelId = "441327731486097429"
+    static let botDevUserId = "290483761559240704"
```
Comment (MahdiBM) [inline CODE/Sources/PennyBOT/Constants.swift:6]: My own user id, which the logs will mention on `error`/`warning` or when the bot restarts.
We can change it to a role too, if you want some other people to be notified as well.

File: CODE/Sources/PennyBOT/CoinService/DefaultCoinService.swift [hunk 4]
```diff
@@ -19,10 +19,10 @@ struct DefaultCoinService: CoinService {
         let data = try JSONEncoder().encode(coinRequest)
         request.body = .bytes(data)
         let response = try await httpClient.execute(request, timeout: .seconds(30), logger: self.logger)
-        logger.trace("HTTP head \(response)")
+        logger.trace("Received HTTP Head", metadata: ["response": "\(response)"])
```
Comment (MahdiBM) [inline CODE/Sources/PennyBOT/CoinService/DefaultCoinService.swift:22]: I've refined log statements to use `metadata`, mainly so they look much nicer in Discord, but they should also look nicer in stdout logs.

File: CODE/Sources/PennyBOT/Penny.swift [hunk 5]
```diff
@@ -52,4 +48,42 @@ struct Penny {
...
+                httpClient: httpClient,
+                token: token,
+                appId: nil
+            ),
+            configuration: .init(
+                fallbackLogger: Logger(
+                    label: "DiscordBMFallback",
+                    factory: StreamLogHandler.standardOutput(label:metadataProvider:)
+                ),
+                aliveNotice: .init(
```
Comment (MahdiBM) [inline CODE/Sources/PennyBOT/Penny.swift:69]: every 3 hours Penny will send a "I'm alive" notice to the logs. This is only if there has been no logs for 3 hours.
also when the bot restarts, Penny will mention me incase it's crashing or something.

File: CODE/Sources/PennyBOT/Penny.swift [hunk 6]
```diff
@@ -52,4 +48,42 @@ struct Penny {
...
+                    address: try! .webhook(.url(webhookUrl)),
+                    interval: .hours(3),
+                    message: "I'm Alive! :)",
+                    initialNoticeMention: .user(Constants.botDevUserId)
+                ),
+                mentions: [
```
Comment (MahdiBM) [inline CODE/Sources/PennyBOT/Penny.swift:75]: mentions the user id (me) on warning and error logs.

File: CODE/Sources/PennyBOT/Penny.swift [hunk 7]
```diff
@@ -52,4 +48,42 @@ struct Penny {
...
+                    .warning: .user(Constants.botDevUserId),
+                    .error: .user(Constants.botDevUserId)
+                ],
+                extraMetadata: [.warning, .error],
```
Comment (MahdiBM) [inline CODE/Sources/PennyBOT/Penny.swift:79]: sends `file`/`function`/`line`/`source` too, on `warning` or `error`.

File: CODE/Sources/PennyBOT/Penny.swift [hunk 8]
```diff
@@ -52,4 +48,42 @@ struct Penny {
...
+            fatalError("Missing 'LOGGING_WEBHOOK_URL' or 'BOT_TOKEN' env vars")
+        }
+        DiscordGlobalConfiguration.logManager = DiscordLogManager(
+            client: DefaultDiscordClient(
+                httpClient: httpClient,
+                token: token,
+                appId: nil
+            ),
+            configuration: .init(
+                fallbackLogger: Logger(
```
Comment (MahdiBM) [inline CODE/Sources/PennyBOT/Penny.swift:65]: If there is a problem with sending logs to Discord, the problems will be reported using this logger.

**Review by MahdiBM — COMMENTED** [review]
Left some comments, mostly just explaining.

File: CODE/task-def.json (hunk 1 above)
Comment (MahdiBM) [reply CODE/task-def.json:39]: Anyone having the link will be able to send logs to the channel just like us. So it's kind-of secret. The url doesn't have any other authorization method, other than being a long string that other people wouldn't be able to guess.

Comment (MahdiBM) [reply CODE/task-def.json:39]: I could just set it to send the logs directly to the channel with normal messages, too. I would prefer the webhook way, but it's not that important.

Comment (0xTim) [reply CODE/task-def.json:39]: Cool, it's no problem. If you set up the Webhook URL and DM me on Discord I can add it to secrets manager and update the ARN

**Review by 0xTim — APPROVED** [review]

**Pull request merged**